### ISSUE TITLE
istio pilot crash on load

### ISSUE THREAD

**Issue by RaynDol** [istio/istio#14537]
(This is used to report product bugs, please visit <https://discuss.istio.io> for questions on using Istio)

**Bug description**
we are using istio latest `1.1.7` using helm default config.
we are doing load test before using istio in production env

in the test we target to create `4500`  pods (during 30 minute), each pod contain 3 routes however 
istio pilot was **crashed** after `2750` pods.  (also istio-ingressgateway & istio-policy)
using prometheus we saw that during the creation of the pods the pilot use `53GB` of memory,...
1. what could be the reason for the high memory (53GB) consumption  ? 
2. should we update any value here to avoid crash of pilot on heavy load?
https://istio.io/docs/reference/config/installation-options/#pilot-options

The error message which the istio pilot was down was
The node was law in resource, memory container discovery was using `31084218ki`, which exceedes  its request for 2GI. 

The pod error was crashloopback ,

Any suggestion/hint how we can overcome this issue ?

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[x] Performance and Scalability
[x] Policies and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience

**Expected behavior**
Istio will be functional on high load
**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
1.1.7
**How was Istio installed?**
Helm 
**Environment where bug was observed (cloud vendor, OS, etc)**
K8S  1.13.6
Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by howardjohn** [comment]
Is there just 1 instance of pilot running? Are all the pods in one service and/or namespace or are they all different?

**Comment by howardjohn** [comment]
53GB is not normal for a deployment of this size. You could just bump up the requests, but there is likely something going wrong causing elevated memory usage.

**Comment by hzxuzhonghu** [comment]
Thanks for your report, 53GB is too much, @RaynDol Can you see how pilot crashed? OOM killed or some panic?

@howardjohn Wonder what's the scale you are testing in google. How many pods and how many services?

**Comment by howardjohn** [comment]
@hzxuzhonghu we have some regularly scheduled tests running at http://grafana.v12.qualistio.org/ (seems prometheus is having issues currently though, need to investigate). That one has 450 services and 1100 pods.

http://grafana.v11.qualistio.org also has 450 services and 1700 pods. Pilot is only using 2.5GB in this case, although it is using sidecar.

@RaynDol it would be useful if you could export the grafana Pilot dashboard during the time period. (top right of grafana ui -> share -> snapshot -> publish to raintank).

**Comment by RaynDol** [comment]
@howardjohn - thanks, we tried with the default config which in 1 - 5 instances , we will increase the auto pilot pilot.autoscaleMax to `10` and double the memory of `pilot.resources.requests.memory`  and will try again , is it sufficient for high load or you suggest  to  increase other properties ?

**Comment by RaynDol** [comment]
@hzxuzhonghu - Thanks, the error is OOM .

**Comment by howardjohn** [comment]
@RaynDol I even just a couple pilots should handle that load easily which is why I think you may be hitting a memory leak or something similar, so raising the limits may just hide the problem

**Comment by RaynDol** [comment]
@howardjohn - currently we dont use side car for each pods i.e. we didnt create `kubectl label namespace tals istio-injection=enabled` for each namespace, is it related somehow ? is it must ? 
btw, there is command which enables it automatically for `all` the namespace instead of configuring it for each space ?

**Comment by howardjohn** [comment]
@rayndol you can set `sidecarInjectorWebhook.enableNamespacesByDefault`.

Not having them all injected should mean less load for pilot (each sidecar == 1 more connection to handle)

**Comment by RaynDol** [comment]
@howardjohn - thanks,

we are using the following `ingress-gateway ` maybe this could be the reason for the hign-memory consumption? should we change something ? 


```
---
# Source: istio/charts/gateways/templates/deployment.yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  name: istio-ingressgateway
  namespace: istio-system
  labels:
    chart: gateways
    heritage: Tiller
    release: istio
    app: istio-ingressgateway
    istio: ingressgateway
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
      istio: ingressgateway
  template:
    metadata:
      labels:
        chart: gateways
        heritage: Tiller
        release: istio
        app: istio-ingressgateway
        istio: ingressgateway
      annotations:
        sidecar.istio.io/inject: "false"
    spec:
      serviceAccountName: istio-ingressgateway-service-account
      containers:
        - name: ingress-sds
          image: "gcr.io/istio-release/node-agent-k8s:master-latest-daily"
          imagePullPolicy: IfNotPresent
          resources:
            limits:
              cpu: 2000m
              memory: 1024Mi
            requests:
              cpu: 100m
              memory: 128Mi
            
          env:
          - name: "ENABLE_WORKLOAD_SDS"
            value: "false"
          - name: "ENABLE_INGRESS_GATEWAY_SDS"
            value: "true"
          - name: "INGRESS_GATEWAY_NAMESPACE"
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
          volumeMounts:
          - name: ingressgatewaysdsudspath
            mountPath: /var/run/ingress_gateway
        - name: istio-proxy
          image: "gcr.io/istio-release/proxyv2:master-latest-daily"
          imagePullPolicy: IfNotPresent
          ports:
            - containerPort: 15020
            - containerPort: 80
            - containerPort: 443
            - containerPort: 31400
            - containerPort: 15029
            - containerPort: 15030
            - containerPort: 15031
            - containerPort: 15032
            - containerPort: 15443
            - containerPort: 15090
              protocol: TCP
              name: http-envoy-prom
          args:
          - proxy
          - router
          - --domain
          - $(POD_NAMESPACE).svc.cluster.local
          - --log_output_level=default:info
          - --drainDuration
          - '45s' #drainDuration
          - --parentShutdownDuration
          - '1m0s' #parentShutdownDuration
          - --connectTimeout
          - '10s' #connectTimeout
          - --serviceCluster
          - istio-ingressgateway
          - --zipkinAddress
          - zipkin:9411
          - --proxyAdminPort
          - "15000"
          - --statusPort
          - "15020"
          - --controlPlaneAuthPolicy
          - NONE
          - --discoveryAddress
          - istio-pilot:15010
          readinessProbe:
            failureThreshold: 30
            httpGet:
              path: /healthz/ready
              port: 15020
              scheme: HTTP
            initialDelaySeconds: 1
            periodSeconds: 2
            successThreshold: 1
            timeoutSeconds: 1
          resources:
            limits:
              cpu: 4000m
              memory: 2024Mi
            requests:
              cpu: 100m
              memory: 128Mi
            
          env:
          - name: NODE_NAME
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: spec.nodeName
          - name: POD_NAME
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.name
          - name: POD_NAMESPACE
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
          - name: INSTANCE_IP
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: status.podIP
          - name: HOST_IP
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: status.hostIP
          - name: ISTIO_META_POD_NAME
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.name
          - name: ISTIO_META_CONFIG_NAMESPACE
            valueFrom:
              fieldRef:
                fieldPath: metadata.namespace
          - name: ISTIO_META_USER_SDS
            value: "true"
          - name: ISTIO_META_ROUTER_MODE
            value: sni-dnat
          volumeMounts:
          - name: ingressgatewaysdsudspath
            mountPath: /var/run/ingress_gateway
          - name: istio-certs
            mountPath: /etc/certs
            readOnly: true
          - name: ingressgateway-certs
            mountPath: "/etc/istio/ingressgateway-certs"
            readOnly: true
          - name: ingressgateway-ca-certs
            mountPath: "/etc/istio/ingressgateway-ca-certs"
            readOnly: true
      volumes:
      - name: ingressgatewaysdsudspath
        emptyDir: {}
      - name: istio-certs
        secret:
          secretName: istio.istio-ingressgateway-service-account
          optional: true
      - name: ingressgateway-certs
        secret:
          secretName: "istio-ingressgateway-certs"
          optional: true
      - name: ingressgateway-ca-certs
        secret:
          secretName: "istio-ingressgateway-ca-certs"
          optional: true
      affinity:      
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - amd64
                - ppc64le
                - s390x
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 2
            preference:
              matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - amd64
          - weight: 2
            preference:
              matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - ppc64le
          - weight: 2
            preference:
              matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - s390x      
---

```

**Comment by RaynDol** [comment]
currently we are blocked, please advice how to proceed

**Comment by RaynDol** [comment]
@howardjohn - I've tried to do port forwarding to get the heap as you suggest via `kubectl -n istio-system port-forward $(kubectl -n istio-system get pod -l app=itsio-pilot -o jsonpath='{.items[0].metadata.name}') 8080:8080` and get error: `error: error executing jsonpath "{.items[0].metadata.name}": Error executing template: array index out of bounds: index 0, length 0. Printing more information for debugging the template:
	template was:
		{.items[0].metadata.name}
	object given to jsonpath engine was:
		map[string]interface {}{"apiVersion":"v1", "items":[]interface {}{}, "kind":"List", "metadata":map[string]interface {}{"resourceVersion":"", "selfLink":""}}
`
Can you please add the full command?

**Comment by RaynDol** [comment]
any idea how to provide the `pprof` data as the port forwarding to pilot is failing...is there any instruction how to provide it?

**Comment by howardjohn** [comment]
You command has a type `itsio-pilot`. You can just do `kubectl -n istio-system port-forward svc/istio-pilot 8080` though

**Comment by RaynDol** [comment]
@howardjohn - when running this command it generates pprof folder with  tar and when extracting it I see `.pb` file which is half empty,`pprof.pilot-discovery.alloc_objects.alloc_space.inuse_objects.inuse_space.001.pb`
Should it be like this?

**Comment by howardjohn** [comment]
I normally do ` pprof -http=:8090 /home/howardjohn/pprof/pprof.pilot-discovery.goroutine.001.pb.gz` which starts a local server you can browse. But as far as sharing it, I think the two options are exporting to svg (something like `go tool pprof -svg profile300.gz` is what I got from https://coder.today/tech/2018-11-10_profiling-your-golang-app-in-3-steps/) or upload it here

**Comment by howardjohn** [comment]
You can also send it to me over slack or email if you are concerned about sensative info, although I don't think it has any

**Comment by RaynDol** [comment]
@howardjohn -

1.  Not sure if this is related to our scenario , but currently we are not using the 'inject` label for each pod as discussed before, is there a way somehow to make pilot (with some config) to reduce the amount of work (maybe there is things which is not necessarily needed for prod) ? 

2. we are using virtual service for each pods , is there a way somehow to provide only one virtual service to rule them all (all our pods)  ?  if yes please provide a reference , I think that for us this may help to solve the issue...

Btw,
which slack channel I should post this, or which email ?

**Comment by howardjohn** [comment]
Slack channel + instructions are on https://istio.io/about/community/join/. My email is my github name `@google.com`.

Using the [Sidecar](https://istio.io/docs/reference/config/networking/v1alpha3/sidecar/) config you can specify what config to send to each envoy. So you can specify that each pod should only get config for services in its own namespace for example.

I don't think consolidating VS will help, internally they will be modeled essentially the same. Either way -- if you are seeing 53GB usage something is going very wrong (memory leak, etc), so i don't think a microoptimization would help much.

**Comment by RaynDol** [comment]
Thanks, As we are **not** using side car yet for all the services I think this wouldn't help us to reduce the pilot memory consumption...

**Comment by howardjohn** [comment]
It still would, say you have namespace X with all injected, but all other namespaces are not yet injected. The proxies in X will still get config for every service. But you can use the Sidecar resource and they will just get config for services in namespace X (or whatever you configure)

**Comment by howardjohn** [comment]
This is probably fixed by #14782. It would be great if you can try the 1.3rc1 and let us know

**Comment by howardjohn** [comment]
Presumed fixed in 1.3 -- please open and issues and mention me if you are still seeing issues

**Comment by erdun** [comment]
In my scene， goroutines and memory increase sharply while there're not many pods creating.

istioctl version
```
client version: 1.6.10
pilot version: 1.5.0
pilot version: 1.5.0
pilot version: 1.5.0
pilot version: 1.5.0
pilot version: 1.5.0
```

![image](https://user-images.githubusercontent.com/16429864/99620425-769a9000-2a60-11eb-9405-938ada8462fd.png)
![image](https://user-images.githubusercontent.com/16429864/99620567-c0837600-2a60-11eb-945b-fec64b68c092.png)
